### PR TITLE
Add TryFromBytes::try_{ref,mut}_from_{bytes,prefix,suffix}_with_elems

### PR DESCRIPTION
**Pull request by joshlf**
Makes progress on #5

### CODE REVIEW

**Review by jswrenn — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -2011,6 +2011,360 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// let src = &mut [85, 85][..];
+    /// let (zsty, _) = ZSTy::try_mut_from_prefix_with_elems(src, 42).unwrap();
+    /// assert_eq!(zsty.trailing_dst.len(), 42);
+    /// ```
+    ///
+    /// [`try_mut_from_prefix`]: TryFromBytes::try_mut_from_prefix
+    #[must_use = "has no side effects"]
+    #[inline]
+    fn try_mut_from_prefix_with_elems(
```
Comment (joshlf) [inline src/lib.rs:2269]: Also need `try_mut_from_bytes_with_elems`.

File: src/lib.rs [hunk 2]
```diff
@@ -2011,6 +2011,360 @@ pub unsafe trait TryFromBytes {
...
+    /// }
+    ///
+    /// let bytes = &mut [0xC0, 0xC0, 240, 77, 2, 3, 4, 5, 6, 7, 8][..];
+    ///
+    /// let (packet, suffix) = Packet::try_mut_from_prefix_with_elems(bytes, 3).unwrap();
+    ///
+    /// assert_eq!(packet.mug_size, 240);
+    /// assert_eq!(packet.temperature, 77);
+    /// assert_eq!(packet.marshmallows, [[2, 3], [4, 5], [6, 7]]);
+    /// assert_eq!(suffix, &[8u8][..]);
```
Comment (joshlf) [inline src/lib.rs:2238]: Demonstrate mutation like we do in other `mut` method examples?

File: src/lib.rs [hunk 3]
```diff
@@ -2011,6 +2011,360 @@ pub unsafe trait TryFromBytes {
...
+    /// }
+    ///
+    /// let bytes = &mut [123, 0xC0, 0xC0, 240, 77, 2, 3, 4, 5, 6, 7][..];
+    ///
+    /// let (prefix, packet) = Packet::try_mut_from_suffix_with_elems(bytes, 3).unwrap();
+    ///
+    /// assert_eq!(packet.mug_size, 240);
+    /// assert_eq!(packet.temperature, 77);
+    /// assert_eq!(packet.marshmallows, [[2, 3], [4, 5], [6, 7]]);
+    /// assert_eq!(prefix, &[123u8][..]);
```
Comment (joshlf) [inline src/lib.rs:2327]: Demonstrate mutation like we do in other `mut` method examples?

File: src/lib.rs (hunk 1 above)
Comment (jswrenn) [reply src/lib.rs:2269]: lol d'oh

File: src/lib.rs (hunk 2 above)
Comment (jswrenn) [reply src/lib.rs:2351]: [Done](https://github.com/google/zerocopy/compare/b0d1b2f4989527757c7fc00e9eb4ba34a0270a83..cb103be9b44c3882c7c92cdcb0f22200129789ac#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R2353-R2357)

File: src/lib.rs (hunk 3 above)
Comment (jswrenn) [reply src/lib.rs:2445]: [Done](https://github.com/google/zerocopy/compare/b0d1b2f4989527757c7fc00e9eb4ba34a0270a83..cb103be9b44c3882c7c92cdcb0f22200129789ac#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R2447-R2451)

File: src/lib.rs (hunk 1 above)
Comment (jswrenn) [reply src/lib.rs:2387]: [Done](https://github.com/google/zerocopy/compare/b0d1b2f4989527757c7fc00e9eb4ba34a0270a83..cb103be9b44c3882c7c92cdcb0f22200129789ac#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R2195-R2303)

**Comment by jswrenn** [comment]
Hit merge when you're satisfied, @joshlf.

File: src/lib.rs [hunk 4]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// [`try_mut_from_bytes`]: TryFromBytes::try_mut_from_bytes
+    #[must_use = "has no side effects"]
+    #[inline]
+    fn try_mut_from_bytes_with_elems(
+        source: &mut [u8],
+        count: usize,
+    ) -> Result<&mut Self, TryCastError<&mut [u8], Self>>
+    where
+        Self: KnownLayout<PointerMetadata = usize> + Immutable,
```
Comment (joshlf) [inline src/lib.rs:2280]: Remove the Immutable bound?

File: src/lib.rs [hunk 5]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// [`try_mut_from_prefix`]: TryFromBytes::try_mut_from_prefix
+    #[must_use = "has no side effects"]
+    #[inline]
+    fn try_mut_from_prefix_with_elems(
+        source: &mut [u8],
+        count: usize,
+    ) -> Result<(&mut Self, &mut [u8]), TryCastError<&mut [u8], Self>>
+    where
+        Self: KnownLayout<PointerMetadata = usize> + Immutable,
```
Comment (joshlf) [inline src/lib.rs:2392]: Remove the Immutable bound?

File: src/lib.rs [hunk 6]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// [`try_mut_from_prefix`]: TryFromBytes::try_mut_from_prefix
+    #[must_use = "has no side effects"]
+    #[inline]
+    fn try_mut_from_suffix_with_elems(
+        source: &mut [u8],
+        count: usize,
+    ) -> Result<(&mut [u8], &mut Self), TryCastError<&mut [u8], Self>>
+    where
+        Self: KnownLayout<PointerMetadata = usize> + Immutable,
```
Comment (joshlf) [inline src/lib.rs:2486]: Remove the Immutable bound?

File: src/lib.rs [hunk 7]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    /// Since an explicit `count` is provided, this method supports types with
+    /// zero-sized trailing slice elements. Methods such as [`try_mut_from_bytes`]
+    /// which do not take an explicit count do not support such types.
+    ///
+    /// ```
+    /// use core::num::NonZeroU16;
+    /// use zerocopy::*;
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(TryFromBytes, Immutable, KnownLayout)]
```
Comment (joshlf) [inline src/lib.rs:2260]: Don't derive Immutable here?

File: src/lib.rs [hunk 8]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![allow(non_camel_case_types)] // For C0::xC0
+    /// use zerocopy::TryFromBytes;
+    /// # use zerocopy_derive::*;
+    ///
+    /// // The only valid value of this type is the byte `0xC0`
+    /// #[derive(TryFromBytes, KnownLayout, Immutable)]
```
Comment (joshlf) [inline src/lib.rs:2216]: Don't derive Immutable here?

File: src/lib.rs [hunk 9]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
```
Comment (joshlf) [inline src/lib.rs:2326]: Don't derive Immutable here?

File: src/lib.rs [hunk 10]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    /// Since an explicit `count` is provided, this method supports types with
+    /// zero-sized trailing slice elements. Methods such as [`try_mut_from_prefix`]
+    /// which do not take an explicit count do not support such types.
+    ///
+    /// ```
+    /// use core::num::NonZeroU16;
+    /// use zerocopy::*;
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(TryFromBytes, Immutable, KnownLayout)]
```
Comment (joshlf) [inline src/lib.rs:2372]: Don't derive Immutable here?

File: src/lib.rs [hunk 11]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![allow(non_camel_case_types)] // For C0::xC0
+    /// use zerocopy::TryFromBytes;
+    /// # use zerocopy_derive::*;
+    ///
+    /// // The only valid value of this type is the byte `0xC0`
+    /// #[derive(TryFromBytes, KnownLayout, Immutable)]
```
Comment (joshlf) [inline src/lib.rs:2420]: Don't derive Immutable here?

File: src/lib.rs [hunk 12]
```diff
@@ -2011,6 +2015,479 @@ pub unsafe trait TryFromBytes {
...
+    /// Since an explicit `count` is provided, this method supports types with
+    /// zero-sized trailing slice elements. Methods such as [`try_mut_from_prefix`]
+    /// which do not take an explicit count do not support such types.
+    ///
+    /// ```
+    /// use core::num::NonZeroU16;
+    /// use zerocopy::*;
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(TryFromBytes, Immutable, KnownLayout)]
```
Comment (joshlf) [inline src/lib.rs:2466]: Don't derive Immutable here?

**Comment by joshlf** [comment]
Awesome, queued!

**Pull request merged**